server/src/main/scala/org/http4s/server/middleware/MaxActiveRequests.scala

```diff
@@ -0,0 +1,55 @@
+package org.http4s.server.middleware
+
+import cats.implicits._
+import cats.data.Kleisli
+import cats.effect._
+import cats.effect.concurrent.Semaphore
+import cats.effect.implicits._
+
+import org.http4s.Status
+import org.http4s.Response
+
+object MaxActiveRequests {
+
+  def byHeaders[F[_]: Concurrent, A](
+    maxActive: Long,
+    defaultResp: Response[F] = Response[F](status = Status.ServiceUnavailable)
+  )(http: Kleisli[F, A, Response[F]]): F[Kleisli[F, A, Response[F]]] = {
+    Semaphore[F](maxActive).map{ sem => 
+      Kleisli{a: A => 
+        sem.tryAcquire.bracket{
+          if (_) http.run(a)
+          else defaultResp.pure[F]
+        }{
+          if (_) sem.release
+          else Sync[F].unit
+        }
+        
+      }
+    }
+  }
+
+  /**
+   * Caveats: Body Of the Response Must Be Run.
```

Comment:
This should be guaranteed by the server, so I'm not so worried about the caveat.  I understand the difference, but I wonder whether `byHeaders` is all that useful.

This is another case where a resource model for the response would be highly useful, though we ran into other friction with that.  I can't remember why we didn't make a `bracketResponse` combinator.

Reply:
Alright, going with the singular interface. Its cleaner by a longshot.

Reply:
Interestingly, the byHeaders actually abstracts over all Kleislis.